### ISSUE TITLE
Swift 5.7 regex literal with imbalanced curly braces breaks linting

### ISSUE THREAD

**Issue by FelixStridsbergPaf** [realm/SwiftLint#5114]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

If you have a regex containing imbalanced curly brackets (e.g: `/{[^}]*}/`) in a `do` block. A pattern matching `catch` with hoisted `let` (i.e. `catch let Foo.bar(aaa) {`) falsely validates the type as an identifier.

Removing the imbalance in the regex, for example `/{[^a]*}/`, or `/{[^{}]*}/` makes the linting pass. Inlining the `let` to `catch Foo.bar(let aaa) {` also passes.

It has a similar trigger as https://github.com/realm/SwiftLint/issues/4610, so maybe related cause.

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint lint
Linting Swift files at paths .
Linting 'test.swift' (1/1)
/test/test.swift:8:11: error: Identifier Name Violation: Variable name 'Foo' should start with a lowercase character (identifier_name)
/test/test.swift:7:12: warning: Opening Brace Spacing Violation: Opening braces should be preceded by a single space and on the same line as the declaration (opening_brace)
Done linting! Found 2 violations, 1 serious in 1 file.
```

### Environment

* SwiftLint version 0.52.4
* Installation method used Homebrew
* Paste your configuration file: None

* Are you using nested configurations: No
* Which Xcode version are you using: Xcode 14.3, Build version 14E222b
* To reproduce:

```swift
enum Foo: Error {
  case bar(aaa: Int)
}

public func test() {
  do {
      _ = /{[^}]*}/ // This triggers the violation on row 8
  } catch let Foo.bar(aaa) { // This is row 8
  } catch {}
}
```

**Comment by SimplyDanny** [comment]
I guess the whole rule needs a rewrite using SwiftSyntax to overcome this and similar issues.
